File: app/models/vote.js
/**
 * The votes
 */
import Basemodel from "./basemodel.js";

/**
 * Represents a vote in the e-voting application.
 * 
 * @class Vote
 * @extends {Basemodel}
 * 
 * @property {string} collection - The name of the collection in the database.
 * @property {string} candidate_id - The ID of the candidate being voted for.
 * @property {string} event_id - The ID of the event where the vote is cast.
 * @property {string} category_id - The ID of the category in which the vote is cast.
 * @property {number} number_of_votes - The number of votes cast (default is 1).
 * @property {string} voter_ip - The IP address of the voter.
 * @property {string} [bundle_id] - The ID of the vote bundle (optional).
 * 
 * @param {string} candidate_id - The ID of the candidate being voted for.
 * @param {string} event_id - The ID of the event where the vote is cast.
 * @param {string} category_id - The ID of the category in which the vote is cast.
 * @param {number} [number_of_votes=1] - The number of votes cast (default is 1).
 * @param {string} voter_ip - The IP address of the voter.
 * @param {string} [bundle_id] - The ID of the vote bundle (optional).
 * @param {...any} kwargs - Additional arguments to be passed to the base model.
 */
class Vote extends Basemodel{

    static collection = "votes";

    constructor(candidate_id, event_id,
                category_id, number_of_votes= 1,
                voter_ip, bundle_id = null, ...kwargs) {
        super(...kwargs);
        this.candidate_id = candidate_id;
        this.event_id = event_id;
        this.category_id = category_id;
        this.number_of_votes = number_of_votes;
        this.voter_ip = voter_ip;
        this.bundle_id = bundle_id;
    }
}

export default Vote;

File: app/models/vote.py
from models.base_model import BaseModel

class Vote(BaseModel):
    collection = "votes"

    def __init__(self, candidate_id, event_id, category_id, voter_ip, number_of_votes=1, **kwargs):
        super().__init__(**kwargs)
        self.candidate_id = candidate_id
        self.event_id = event_id
        self.category_id = category_id
        self.number_of_votes = number_of_votes
        self.voter_ip = voter_ip

class VoteBundle(BaseModel):
    collection = "vote_bundles"

    def __init__(
        self,
        name,
        category_id,
        price_per_vote,
        bundle_size=1,
        discount=0.0,
        promo_code=None,
        promo_discount=0.0,
        is_active=True,
        **kwargs
    ):
        super().__init__(**kwargs)
        self.name = name  # e.g., "Standard", "Promo Pack"
        self.category_id = category_id
        self.price_per_vote = price_per_vote
        self.bundle_size = bundle_size
        self.discount = discount  # e.g., 0.1 for 10% off
        self.promo_code = promo_code
        self.promo_discount = promo_discount  # e.g., 0.2 for 20% off with promo
        self.is_active = is_active

    def to_dict(self):
        return {
            "id": getattr(self, "id", None),
            "name": self.name,
            "category_id": self.category_id,
            "price_per_vote": self.price_per_vote,
            "bundle_size": self.bundle_size,
            "discount": self.discount,
            "promo_code": self.promo_code,
            "promo_discount": self.promo_discount,
            "is_active": self.is_active,
        }

    @classmethod
    def get_active_bundles_by_category(cls, category_id):
        # Returns all active bundles for a given category
        return cls.all({"category_id": category_id, "is_active": True})

    @classmethod
    def get_all_bundles_by_category(cls, category_id):
        return cls.all({"category_id": category_id})
